src/System.Net.Security/tests/FunctionalTests/KerberosTest.cs

```diff
@@ -0,0 +1,474 @@
+                Assert.True(_secondMessage.SequenceEqual(secondRecvBuffer), "second message received is as expected");
+                finished = svrMsgTask.Wait(TestConfiguration.PassingTestTimeoutMilliseconds);
+                Assert.True(finished, "Message roundtrip completed in the allotted time");
+            }
+        }
+
+        [Fact, OuterLoop]
+        [PlatformSpecific(PlatformID.Linux)]
+        public void NegotiateStream_EchoServer_ClientWriteRead_Successive_Async_Success()
+        {
+            if (!_isKrbAvailable)
+            {
+                return;
+            }
+
+            VirtualNetwork network = new VirtualNetwork();
+            byte[] firstRecvBuffer = new byte[_firstMessage.Length];
+            byte[] secondRecvBuffer = new byte[_secondMessage.Length];
+
+            using (var clientStream = new VirtualNetworkStream(network, isServer: false))
+            using (var serverStream = new VirtualNetworkStream(network, isServer: true))
+            using (var client = new NegotiateStream(clientStream))
+            using (var server = new UnixGssFakeNegotiateStream(serverStream))
+            {
+                Assert.False(client.IsAuthenticated, "client is not authenticated before AuthenticateAsClient call");
+
+                Task[] auth = new Task[2];
+                string user = string.Format("{0}@{1}", TestConfiguration.KerberosUser, TestConfiguration.Realm);
+                string target = string.Format("{0}@{1}", TestConfiguration.HostTarget, TestConfiguration.Realm);
+                NetworkCredential credential = new NetworkCredential(user, _fixture.password);
+                auth[0] = client.AuthenticateAsClientAsync(credential, target);
+                auth[1] = server.AuthenticateAsServerAsync();
+                bool finished = Task.WaitAll(auth, TestConfiguration.PassingTestTimeoutMilliseconds);
+                Assert.True(finished, "Handshake completed in the allotted time");
+
+                Task serverTask = server.PollMessageAsync(2);
+                Task[] msgTasks = new Task[5];
+                msgTasks[0] = client.WriteAsync(_firstMessage, 0, _firstMessage.Length);
+                msgTasks[1] = client.WriteAsync(_secondMessage, 0, _secondMessage.Length);
+                msgTasks[2] = client.ReadAsync(firstRecvBuffer, 0, firstRecvBuffer.Length);
+                msgTasks[3] = client.ReadAsync(secondRecvBuffer, 0, secondRecvBuffer.Length);
+                msgTasks[4] = serverTask;
+                finished = Task.WaitAll(msgTasks, TestConfiguration.PassingTestTimeoutMilliseconds);
+                Assert.True(finished, "Messages sent and received in the allotted time");
+                Assert.True(_firstMessage.SequenceEqual(firstRecvBuffer), "The first message received is as expected");
+                Assert.True(_secondMessage.SequenceEqual(secondRecvBuffer), "The second message received is as expected");
```

Comment:
Is your order guaranteed here?  You called two writes, and two reads, but since the second write doesn't require the first to have finished (such as being a ContinueWith) they could, possibly, be sent (or read) in reverse... no?